# MiCM Workshop Series
# R - Beyond the Basics
## Efficient Coding
### - Yi Lian
### - March 6, 2020
__Link to workshop material__  https://github.com/ly129/MiCM2020

<a id='0'></a>
## Outline
1. [__An overview of efficiency__](#1)
    - General rules
    - R-specific rules
    - R object types (if necessary)
    - Record runtime of your code
2. [__Efficient coding__](#2)
    - Powerful functions in R
            - aggregate(), by(), apply() family
            - ifelse(), cut() and split()
    - Write our own functions in R
            - function()
    - Examples
         - Categorization, conditional operations, etc..
3. [__Efficient computing__](#3) (not covered today)
    - Parallel computing
    - Rcpp
    - Fortran
3. [__Exercises__](#4)

##### Important note! There are MANY advanced and powerful packages that do different things. There are too many and they are too diverse to be covered in this workshop.
___Here is a list of some awesome packages.___ https://awesome-r.com/

### 1.1 General rules
### 1.2 R-specific rules

### 1.3 R data types and structures
#### 1.3.1 R data types
    - numeric
        - integer
        - double precision (default)
    - logical
    - character
    - factor
    - ...

In [1]:
# double
class(5); is.double(5)

[1] "numeric"

[1] TRUE

In [2]:
# integer
class(5L); is.double(5L)

[1] "integer"

[1] FALSE

In [3]:
# How precise is double precision?
options(digits = 22) # show more digits in output
print(1/3)
options(digits = 7) # back to the default

[1] 0.3333333333333333148296


In [4]:
object.size(rep(5, 10))
object.size(rep(5L, 10))

176 bytes

96 bytes

In [5]:
# logical
class(TRUE); class(F)

[1] "logical"

[1] "logical"

In [6]:
# character
class("TRUE")

[1] "character"

In [7]:
# Not important for this workshop
fac <- as.factor(c(1, 5, 11, 3))
fac

[1] 1  5  11 3 
Levels: 1 3 5 11

In [8]:
class(fac)

[1] "factor"

In [9]:
# R has an algorithm to decide the order of the levels
fac.ch <- as.factor(c("B", "a", "1", "ab", "b", "A"))
fac.ch

[1] B  a  1  ab b  A 
Levels: 1 a A ab b B

#### 1.3.2 R data structures
    - Scalar *
    - Vector
    - Matrix
    - Array
    - List
    - Data frame
    - ...

In [10]:
# Scalar - a vector of length 1
myscalar <- 5
myscalar

[1] 5

In [11]:
class(myscalar)

[1] "numeric"

In [12]:
# Vector
myvector <- c(1, 1, 2, 3, 5, 8)
myvector

[1] 1 1 2 3 5 8

In [13]:
class(myvector)

[1] "numeric"

In [14]:
# Matrix - a 2d array
mymatrix <- matrix(c(1, 1, 2, 3, 5, 8), nrow = 2, byrow = FALSE)
mymatrix

[,1] [,2] [,3]
[1,] 1    2    5   
[2,] 1    3    8

In [15]:
class(mymatrix)

[1] "matrix"

In [16]:
str(mymatrix)

 num [1:2, 1:3] 1 1 2 3 5 8


In [17]:
# Array - not important for this workshop
myarray <- array(c(1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144), dim = c(2, 2, 3))
print(myarray) # print() is not needed if run in R or Rstudio.

, , 1

     [,1] [,2]
[1,]    1    2
[2,]    1    3

, , 2

     [,1] [,2]
[1,]    5   13
[2,]    8   21

, , 3

     [,1] [,2]
[1,]   34   89
[2,]   55  144



In [18]:
class(myarray)

[1] "array"

In [19]:
# List - very important for the workshop
mylist <- list(Title = "R Beyond the Basics",
               Duration = c(2, 2),
               sections = as.factor(c(1, 2, 3, 4)),
               Date = as.Date("2020-03-06"),
               Lunch_provided = FALSE,
               Feedbacks = c("Amazing!", "Great workshop!", "Yi is the best!", "Wow!")
)
print(mylist) # No need for print if running in R or Rstudio

$Title
[1] "R Beyond the Basics"

$Duration
[1] 2 2

$sections
[1] 1 2 3 4
Levels: 1 2 3 4

$Date
[1] "2020-03-06"

$Lunch_provided
[1] FALSE

$Feedbacks
[1] "Amazing!"        "Great workshop!" "Yi is the best!" "Wow!"           



In [20]:
class(mylist)

[1] "list"

In [21]:
# Access data stored in lists
mylist$Title

[1] "R Beyond the Basics"

In [22]:
# or
mylist[[6]]

[1] "Amazing!"        "Great workshop!" "Yi is the best!" "Wow!"

In [23]:
# Further
mylist$Duration[1]
mylist[[6]][2]

[1] 2

[1] "Great workshop!"

In [24]:
# Elements in lists can have different data types
lapply(mylist, class) # We will talk about lapply() later

$Title
[1] "character"

$Duration
[1] "numeric"

$sections
[1] "factor"

$Date
[1] "Date"

$Lunch_provided
[1] "logical"

$Feedbacks
[1] "character"

In [25]:
# Elements in list can have different lengths
lapply(mylist, length)

$Title
[1] 1

$Duration
[1] 2

$sections
[1] 4

$Date
[1] 1

$Lunch_provided
[1] 1

$Feedbacks
[1] 4

In [26]:
# Data frames - most commonly used for analyses
head(mtcars)

mpg  cyl disp hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160  110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160  110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108   93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360  175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225  105 2.76 3.460 20.22 1  0  3    1

In [27]:
# Access a column (variable) in data frames
mtcars$mpg

[1] 21.0 21.0 22.8 21.4 18.7 18.1 14.3 24.4 22.8 19.2 17.8 16.4 17.3 15.2 10.4
[16] 10.4 14.7 32.4 30.4 33.9 21.5 15.5 15.2 13.3 19.2 27.3 26.0 30.4 15.8 19.7
[31] 15.0 21.4

### 1.4 Time your program in R
___Illustrations of R rules for efficiency.___
        - proc.time(), system.time()
        - microbenchmark

#### 1.4.1 Vectorized operation vs. loop
__Example__
Calculate the square root of 1 to 1,000,000 using vectorized operation vs. using a for loop.

In [28]:
# Vectorized operation
# system.time(operation)  returns the time needed to run the 'operation'
t <- system.time( x1 <- sqrt(1:1000000) )
head(x1)

[1] 1.000000 1.414214 1.732051 2.000000 2.236068 2.449490

In [29]:
# For loop
x2 <- rep(NA, 1000000)
t0 <- proc.time()
for (i in 1:1000000) {
    x2[i] <- sqrt(i)
}
t1 <- proc.time()

identical(x1, x2) # Check whether results are the same

[1] TRUE

In [30]:
# As we can see, R is not very fast with loops.
t; t1 - t0
# ?proc.time

   user  system elapsed 
  0.006   0.004   0.010 

   user  system elapsed 
  0.069   0.004   0.074 

##### Take-home message
- Use vectorized operations rather than loops for speed in R.
- Loops are more intuitive though.
- Balance between
    - speed
    - your need for speed
    - your level of comfortableness with linear algebra
    - your level of laziness
    - your typing speed
    - ...
- Based on what you are doing
    - dealing with big dataset and expensive calculations?
    - running the code only once or potentially many many times?

#### 1.4.2 Use established functions

__Example__
Calculate the square root using sqrt( ) vs. our own implementation.

In [31]:
# microbenchmark runs the code multiple times and take a summary
# Use well-developped R function
library(microbenchmark)
result <- microbenchmark(sqrt(500),
                         500^0.5,
                         unit = "ns", times = 1000
                        )
summary(result)
# Result in nanoseconds

expr      min lq  mean    median uq  max  neval
1 sqrt(500)  80  90 101.134  94    101 3164 1000 
2 500^0.5   155 165 178.876 171    176 4167 1000

##### In summary, keep the rules in mind, know what you want to do, test your program, time your program.

<a id='2'></a>
## 2. [Efficient coding](#0)
R has many powerful and useful functions that we can use to achieve efficient coding and computing.
### 2.1 Powerful functions in R
##### Let's play with some data.

In [32]:
data <- read.csv("https://raw.githubusercontent.com/ly129/MiCM2020/master/sample.csv", header = TRUE)
head(data, 8)

X Sex    Wr.Hnd NW.Hnd Pulse Smoke Height Age   
1 1 Male   21.4   21.0   63    Never 180.00 19.000
2 2 Male   19.5   19.4   79    Never 165.00 18.083
3 3 Female 16.3   16.2   44    Regul 152.40 23.500
4 4 Female 15.9   16.5   99    Never 167.64 17.333
5 5 Male   19.3   19.4   55    Never 180.34 19.833
6 6 Male   18.5   18.5   48    Never 167.00 22.333
7 7 Female 17.5   17.0   85    Heavy 163.00 17.667
8 8 Male   19.8   20.0   NA    Never 180.00 17.417

In [33]:
summary(data)

       X              Sex         Wr.Hnd          NW.Hnd          Pulse       
 Min.   :  1.00   Female:47   Min.   :13.00   Min.   :12.50   Min.   : 40.00  
 1st Qu.: 25.75   Male  :53   1st Qu.:17.50   1st Qu.:17.45   1st Qu.: 50.25  
 Median : 50.50               Median :18.50   Median :18.50   Median : 71.50  
 Mean   : 50.50               Mean   :18.43   Mean   :18.39   Mean   : 69.90  
 3rd Qu.: 75.25               3rd Qu.:19.50   3rd Qu.:19.52   3rd Qu.: 84.75  
 Max.   :100.00               Max.   :23.20   Max.   :23.30   Max.   :104.00  
                                                              NA's   :6       
   Smoke        Height           Age       
 Heavy: 6   Min.   :152.0   Min.   :16.92  
 Never:79   1st Qu.:166.4   1st Qu.:17.58  
 Occas: 5   Median :170.2   Median :18.46  
 Regul:10   Mean   :171.8   Mean   :20.97  
            3rd Qu.:179.1   3rd Qu.:20.21  
            Max.   :200.0   Max.   :73.00  
            NA's   :13                     

#### a1. Calculate the mean writing hand span of all individuals
    mean(x, trim = 0, na.rm = FALSE, ...)

#### a2. Calculate the mean height of all individuals, exclude the missing values

#### a3. Calculate the mean of all continuous variables
    apply(X, MARGIN, FUN, ...)

In [34]:
# Choose the continuous variables

In [35]:
# Calculate the mean

#### b1. Calculate the count/proportion of females and males
    table(...,
      exclude = if (useNA == "no") c(NA, NaN),
      useNA = c("no", "ifany", "always"),
      dnn = list.names(...), deparse.level = 1)

    prop.table()

#### b2. Calculate the count in each Smoke group

#### b3. Calculate the count of males and females in each Smoke group

#### c1. Calculate the standard deviation of writing hand span of females
        aggregate()
        tapply()
        by()

In [36]:
# aggregate() syntax 1

In [37]:
# aggregate() syntax 2

In [38]:
# by()

In [39]:
# tapply()

In [40]:
# Return a list using tapply()


##### aggregate( ), by( ) and tapply( ) are all connected. They give different types of output.

#### c2. Calculate the standard deviation of writing hand span of all different Sex-Smoke groups

#### c3. Calculate the standard deviation of writing hand and non-writing hand span of all Sex-Smoke groups

##### Let's try to figure out what aggregate( ) is doing
        print()

##### Exercise.
1. Repeat b1-b3 using aggregate( )

2. Make histograms of writing hand span for all eight Sex-Smoke groups using aggregate( )
        hist()

#### d1. Categorize 'Age' - make a new binary variable 'Adult'
    ifelse(test, yes, no)

In [41]:
vec <- 1:5
vec

ifelse(vec>3, yes = "big", no = "small")

[1] 1 2 3 4 5

[1] "small" "small" "small" "big"   "big"

##### R has if (test) {opt1} else {opt2}, what is the advantage of ifelse( )?

In [42]:
if (data$Age >= 18) {
    data$Adult2 = "Yes"
} else {
    data$Adult2 = "No"
}
head(data)

Warning message in if (data$Age >= 18) {:
“the condition has length > 1 and only the first element will be used”

X Sex    Wr.Hnd NW.Hnd Pulse Smoke Height Age    Adult2
1 1 Male   21.4   21.0   63    Never 180.00 19.000 Yes   
2 2 Male   19.5   19.4   79    Never 165.00 18.083 Yes   
3 3 Female 16.3   16.2   44    Regul 152.40 23.500 Yes   
4 4 Female 15.9   16.5   99    Never 167.64 17.333 Yes   
5 5 Male   19.3   19.4   55    Never 180.34 19.833 Yes   
6 6 Male   18.5   18.5   48    Never 167.00 22.333 Yes

In [43]:
# Delete Adult2
data <- subset(data, select=-c(Adult2))

##### ifelse( ) is vectorized!!!
#### d2. Categorize 'Wr.Hnd' into 5 groups - make a new categorical variable with 5 levels
    1. =< 16: TP/XS
    2. 16~18: P/S
    3. 18~20: M/M
    4. 20~22: G/L
    5. >  22: TG/XL
Can we still use ifelse( )?
    
    cut(x, breaks, labels = NULL, right = TRUE, ...)

In [44]:
cut.points <- c(0, 16, 18, 20, 22, Inf)

head(data)
# labels as default

X Sex    Wr.Hnd NW.Hnd Pulse Smoke Height Age   
1 1 Male   21.4   21.0   63    Never 180.00 19.000
2 2 Male   19.5   19.4   79    Never 165.00 18.083
3 3 Female 16.3   16.2   44    Regul 152.40 23.500
4 4 Female 15.9   16.5   99    Never 167.64 17.333
5 5 Male   19.3   19.4   55    Never 180.34 19.833
6 6 Male   18.5   18.5   48    Never 167.00 22.333

In [45]:
# Set labels to false


In [46]:
# Customized labels
label <- c("TP/XS", "P/S", "M/M", "G/L", "TG/XL")


#### e1. Calculate the mean Wr.Hnd span of each Hnd.group

#### e2. Calcuate the mean Wr.Hnd span of each Hnd.group without using aggregate, by, tapply
        split(x, f, ...)
        lapply(X, FUN, ...)
        sapply(X, FUN, ..., simplify = TRUE)

In [47]:
# cut.points <- c(0, 16, 18, 20, 22, Inf)


In [48]:
# lapply


In [49]:
# sapply


In [50]:
# vapply *
# Safer than sapply(), and a little bit faster
# because FUN.VALUE has to be specified that length and type should match

# va <- vapply(Wr.Hnd.Grp, summary, FUN.VALUE = c("Min." = numeric(1),
#                                                 "1st Qu." = numeric(1),
#                                                 "Median" = numeric(1),
#                                                 "Mean" = numeric(1),
#                                                 "3rd Qu." = numeric(1),
#                                                 "Max." = numeric(1)))
# va

#### f. Calculate the 95% sample confidence intervals of Wr.Hnd in each Smoke group.
One variable for lower bound and one variable for upper bound.

$$ CI = \bar{x} \pm t_{n-1, 0.025} \times \sqrt{\frac{s^2}{n}} $$

where $\bar{x}$ is the sample mean and $s^2$ is the sample variance.

In [51]:
# aggregate(Wr.Hnd~Smoke, data = data, FUN = ...)
# tapply(X = data$Wr.Hnd, INDEX = list(data$Smoke), FUN = ...)

##### Unfortunately, I do not know any function in R that does this calculation.
But we know how to do it step by step.

In [52]:

# How many times did we aggregate according to the group? Can on aggregate only once?

##### Or, we can make our own function and integrate it into aggregate( ), by( ), or tapply( ) !!!
### 2.2 Write our own functions in R
A function takes in some inputs and gives some outputs

In [53]:
# The structure
func_name <- function(argument){
    statement
}

#### Example 1. Make a function for $f(x) = 2x$

In [54]:
# Build the function
times2 <- function(x) {
    fx = 2 * x
    return(fx)
}
# Use the function
times2(x = 5)
# or
times2(3)

[1] 10

[1] 6

#### Example 2. Make a function to calculate the integer division of $a$ by $b$, return the integer part and the modulus.

In [55]:
# R has operators that do this
9 %/% 2
9 %% 2

[1] 4

[1] 1

     floor( ) takes a single numeric argument x and returns a numeric vector containing the largest integers not greater than the corresponding elements of x.



In [56]:
int.div <- function(a, b){
    int <- floor(a/b)
    mod <- a - int*b
    return(list(integer = int, modulus = mod))
}

In [57]:
# class(result)
# Recall: how do we access the modulus?
result <- int.div(21, 4)
result$integer

[1] 5

In [58]:
int.div <- function(a, b){
    int <- a%/%b
    mod <- a%%b
    return(cat(a, "%%", b, ": \n integer =", int,"\n ------------------", " \n modulus =", mod, "\n"))
}
int.div(21,4)

21 %% 4 : 
 integer = 5 
 ------------------  
 modulus = 1 


In [59]:
int.div <- function(a, b){
    int <- a%/%b
    mod <- a%%b
    return(c(a, b))
}
int.div(21, 4)

[1] 21  4

#### Example 3. Make the simplest canadian AI chatbot
A function can return something other than an R object, say some voice.

In [60]:
# No need to worry about the details here.
# Just want to show that functions do not always have to return() something.
AIcanadian <- function(who, reply_to) {
    system(paste("say -v", who, "Sorry!"))
}
AIcanadian("Alex", "Sorry I stepped on your foot.")

In [61]:
# Train my chatbot - AlphaGo style.
# I'll let Alex and Victoria talk to each other.
# MacOS has their voices recorded.
chat_log <- rep(NA, 8)
# for (i in 1:8) {
#     if (i == 1) {
#         chat_log[1] <- "Sorry I stepped on your foot."
#         system("say -v Victoria Sorry, I stepped on your foot.")
#     } else {
#         if (i %% 2 == 0)
#             chat_log[i] <- AIcanadian("Alex", chat_log[i - 1])
#         else
#             chat_log[i] <- AIcanadian("Victoria", chat_log[i - 1])
#     }
# }
# chat_log

#### Example 4. Check one summary statistic by Smoke group of our 'data' data.
Function arguments can be basically anything, say another function.

In [62]:
data_summary <- function(func) {
    data <- read.csv("https://raw.githubusercontent.com/ly129/MiCM2020/master/sample.csv", header = TRUE)
    by(data = data$Wr.Hnd, INDICES = list(data$Smoke), FUN = func)
}
data_summary(mean)

: Heavy
[1] 18.43333
------------------------------------------------------------ 
: Never
[1] 18.31899
------------------------------------------------------------ 
: Occas
[1] 18.44
------------------------------------------------------------ 
: Regul
[1] 19.3

#### Example 5. Default argument value & stop execution

In [63]:
a_times2_unless_you_want.something.else.but.I.refuse.3 <- function(a, b=2){
    if (b == 3) {
        stop("I refuse 3!")
    }
    a*b
}

In [64]:
a_times2_unless_you_want.something.else.but.I.refuse.3(a = 5)

[1] 10

In [65]:
a_times2_unless_you_want.something.else.but.I.refuse.3(a = 5, b = 4)

[1] 20

In [66]:
# a_times2_unless_you_want.something.else.but.I.refuse.3(a = 5, b = 3)

##### Exercise: 
1. Make a function to calculate sample confidence intervals (2.1 f)

2. Use the function in 1 with aggregate(), by() or apply() to calculate the sample confidence intervals (2.1 f)

<a id='3'></a>
## 3. [Efficient computing](#0)
We often want to minimize the resources used to do certain computation.
##### Time is usually the most important resource.
Other resources are relatively less important.
### 3.1 Parallel computing
When multiple tasks are independent of each other. We can use (up to) all the CPU cores at the same time to do the tasks simultaneously. Check CPU usage if interested.

In [67]:
library(parallel)
detectCores()

[1] 4

In [68]:
mat.list <- sapply(c(1, 5, 200, 250, 1800, 2000), diag)
print(head(mat.list, 2)) # print() makes the output here look the same as in R/Rstudio

[[1]]
     [,1]
[1,]    1

[[2]]
     [,1] [,2] [,3] [,4] [,5]
[1,]    1    0    0    0    0
[2,]    0    1    0    0    0
[3,]    0    0    1    0    0
[4,]    0    0    0    1    0
[5,]    0    0    0    0    1



#### Here we compare lapply( ) and its multi-core version mclapply( ) and parLapply( ) in the 'parallel' package.
    - lapply()
    - mclapply(..., mc.preschedule = TRUE)   # without load balancing
    - mclapply(..., mc.preschedule = FALSE)  # with load balancing
    
_mcapply( ) sets up a pool of mc.cores workers just for this computation_

_mclapply( ) is not available on Windows. For those of you using Windows computers_

https://www.apple.com/ca/mac/ or https://ubuntu.com or

    - parLapply                              # without load balancing
    - parLapplyLB                            # with load balancing

To use parLapply( ), we need to set up a cluster, and we need to close the cluster after we are done. The good part is that we can put several parLapply( ) calls within the cluster.

In [69]:
system.time(
    sc <- lapply(mat.list, solve)
)

   user  system elapsed 
  3.627   0.104   4.145 

In [70]:
system.time(
    mc <- mclapply(mat.list, solve, mc.preschedule = TRUE, mc.cores = 3)
)

   user  system elapsed 
  2.054   0.189   3.378 

In [71]:
system.time(
    mc <- mclapply(mat.list, solve, mc.preschedule = FALSE, mc.cores = 3)
)

   user  system elapsed 
  1.291   0.180   1.863 

In [72]:
t <- proc.time()
cl <- makeCluster(3) # Use 3 cores
pl <- parLapply(cl = cl, X = mat.list, fun = solve)
stopCluster(cl)
proc.time() - t

   user  system elapsed 
  0.501   0.109   4.897 

In [73]:
t <- proc.time()
cl <- makeCluster(3)
pl <- parLapplyLB(cl = cl, X = mat.list, fun = solve)
stopCluster(cl)
proc.time() - t

   user  system elapsed 
  0.484   0.094   3.042 

In [74]:
# Two parallel calls within one cluster.
t <- proc.time()
cl <- makeCluster(3)
pl_nb <- parLapply(cl = cl, X = mat.list, fun = solve)
pl_lb <- parLapplyLB(cl = cl, X = mat.list, fun = solve)
stopCluster(cl)
proc.time() - t
# This takes shorter than the sum of the previous two. Why?

   user  system elapsed 
  0.927   0.152   7.207 

##### For both mclapply( ) and parLapply( ), setting up parallel computing takes time (overhead).

In [75]:
t <- proc.time()
cl <- makeCluster(3)
stopCluster(cl)
proc.time() - t

   user  system elapsed 
  0.011   0.012   0.913 

##### Load-balancing is tricky.
"Load balancing is potentially advantageous when the tasks take quite dissimilar amounts of computation time, or where the nodes are of disparate capabilities."

If 1000 tasks need to be allocated to 10 nodes (CPUs, cores, etc.)
- without load-balancing, 100 tasks are sent to each of the nodes.
- with load-balancing, tasks are sent to a node one at a time. Overhead is high.

##### Take-home message
- Parallel computing exists in R
- They are not faster than non-parallel computing by a factor of number of cores used.
- R is still slow.

##### The solution is lower-level programming languages.
- C
- C++
- Fortran

### 3.2 Integration with C++
__The 'Rcpp' package__
    - "Seamless R and C++ Integration"
    
1. Install 'Rcpp' package in R
2. Install compiler
    - Windows: Rstudio should ask you to install Rtools when you source your cpp code.
        - If not, you can download and install Rtools on the exact same webpage where you downloaded R.
    - MacOS:
        - Install XCode Command Line Tools. Open Terminal, paste and run
                xcode-select --install
        - Install gfortran-6.1 binary and clang compiler (also on the exact same webpage as the R download)
        
            https://cran.r-project.org/bin/macosx/tools/
            
    _“Setup is extra work on macOS, but it is above our pay grade to change that.” - Dirk Eddelbuettel_
         https://github.com/RcppCore/RcppArmadillo/issues/249
3. File -> New File -> C++ File
4. Code C++
    - Try not to forget ';' at the end of lines;
    - Every object that has ever appeared has to be defined;
    - Have to use loops to do a lot of things such as matrix calculations (not slow though);

#### Example 1. Create an R function that calculates the square root of vectors in C++.

In [76]:
library(Rcpp)
sourceCpp("sqrt_cpp.cpp")
square_root(1:4)
# We return a NumericVector in the .cpp file. So we get an R vector.

[1] 1.000000 1.414214 1.732051 2.000000

__The addition of the 'RcppArmadillo' package__
      - "Armadillo is a C++ linear algebra library aiming towards a good balance between speed and ease of use."
http://arma.sourceforge.net

___Linear algebra in Armadillo___
http://arma.sourceforge.net/armadillo_joss_2016.pdf

In base C++, operations like matrix multiplication requires loops.
#### Example 2. Create an R function that calculates matrix multiplication in C++.

In [77]:
sourceCpp("mm_cpp.cpp")

In [78]:
# Now we can call the function using the name defined in the .cpp file
set.seed(20190813)
a <- matrix(rnorm(100000), ncol = 50000)  # 2 x 50000 matrix
b <- matrix(rnorm(200000), nrow = 50000)  # 50000 x 4 matrix

mat_mul(a, b)
# We return an Rcpp::List in the .cpp file. So we get an R list here.
# mat_mul(b, a)

$MatrixMultiplication
          [,1]     [,2]       [,3]      [,4]
[1,] -345.3068 359.6366  -54.33261 -182.1485
[2,] -190.4709  85.7216 -330.53902  121.1807

$rows
[1] 2

$cols
[1] 4

In [79]:
bchmk <- microbenchmark(a %*% b,
                        mat_mul(a, b),
                        unit = "us", times = 100
                       )
summary(bchmk)

expr          min     lq       mean     median   uq       max      neval
1 a %*% b       764.153 800.3040 848.7641 826.3525 873.7580 1245.407 100  
2 mat_mul(a, b) 595.822 611.4785 662.4105 630.7145 660.8175 2430.865 100

In [80]:
# Here we make an R function that calls the C++ function
mmc <- function(a, b) {
    result <- mat_mul(a, b)$MatrixMultiplication
    return(result)
}
mmc(a, b)

[,1]      [,2]     [,3]       [,4]     
[1,] -345.3068 359.6366  -54.33261 -182.1485
[2,] -190.4709  85.7216 -330.53902  121.1807

In [81]:
# Another way to do this. Here you do not need to have a separate .cpp file.
# A naive .cpp function is made here.
library(RcppArmadillo)
cppFunction(depends = "RcppArmadillo",
            code = 'arma::mat mm(arma::mat& A, arma::mat& B){
                        return A * B;
                    }'
)

In [82]:
mm(a, b)
# mm(b, a)

[,1]      [,2]     [,3]       [,4]     
[1,] -345.3068 359.6366  -54.33261 -182.1485
[2,] -190.4709  85.7216 -330.53902  121.1807

In [83]:
# We can wrap this naive function in an R function to manipulate input and output in R
mmc2 <- function(A, B) {
    if (ncol(A) == nrow(B)) {
        return(mm(A, B))
    } else {
        stop("non-conformable arguments")
    }
}
mmc2(a, b)
# mmc2(b, a)

[,1]      [,2]     [,3]       [,4]     
[1,] -345.3068 359.6366  -54.33261 -182.1485
[2,] -190.4709  85.7216 -330.53902  121.1807

### 3.3 Integration with Fortran
- Old but even faster
    - We will see that Fortran sacrifices a LOT for speed.
- Require a compiler too
    - MacOS
        - You might have noticed that we have installed a compiler called gfortran
    - Windows
        - A lot of work. See "Fortran_Setup_Win.txt". https://github.com/ly129/MiCM/blob/master/Fortran_Setup_Win.txt
- R can call fortran subroutines
    - Basically a kind of function
    - Through a .so file (Shared Object, MacOS) and a .dll file (Dynamic Link Library, Windows)
        - MacOS Terminal
                cd file_path
                R CMD SHLIB -o file_name.so file_name.f90
        - Windows Command Prompt
                cd file_path
                gfortran -shared -o file_name.dll file_name.f90
- Requires pointers for communication between programs
    - Pointers point to a locations on the computer's memory
        - Two different programs cannot share data directly.
    - The same applies to C++. 'Rcpp' package handles it for us.
    - Test your program in Fortran. The communication between programs make it hard to debug.
        - Make a Fortran program. Then in terminal/command prompt, type
                cd file_path
                gfortran file_name.f90
                ./a.out (MacOS) | a.exe (Windows)

In [84]:
set.seed(20190813)

ra <- 2
ca <- 4
rb <- 4
cb <- 3

A <- matrix(rnorm(ra*ca), nrow = ra)
B <- matrix(rnorm(rb*cb), nrow = rb)

A; B

[,1]       [,2]          [,3]        [,4]      
[1,] -0.7265744  0.3488403409 -1.56818199 0.47863529
[2,] -1.0882637 -0.0002502522  0.03957236 0.01071728

[,1]       [,2]       [,3]      
[1,] -3.1124977  1.1331331 -0.1618750
[2,] -0.4561366 -0.6835550  0.6946241
[3,] -1.5031855  0.1793388 -0.5950162
[4,]  0.5016549  0.3443689 -0.1799074

In [85]:
# Load the executable .so file (MacOS) or .dll file (Windows)
dyn.load("mm_for.so")

In [86]:
# Check whether the "mat_mul_for" function is loaded into R
is.loaded("mat_mul_for")

[1] TRUE

In [87]:
result <- .Fortran("mat_mul_for",
                   A = as.double(A),
                   B = as.double(B),
                   AB = double(ra * cb),  # note the difference here
                   RowA = as.integer(ra),
                   ColA = as.integer(ca),
                   RowB = as.integer(rb),
                   ColB = as.integer(cb),
                   RowAB = as.integer(ra),
                   ColAB = as.integer(cb)
)
result
class(result)

$A
[1] -0.7265743596 -1.0882637317  0.3488403409 -0.0002502522 -1.5681819905
[6]  0.0395723591  0.4786352927  0.0107172815

$B
 [1] -3.1124977 -0.4561366 -1.5031855  0.5016549  1.1331331 -0.6835550
 [7]  0.1793388  0.3443689 -0.1618750  0.6946241 -0.5950162 -0.1799074

$AB
[1]  4.6997204  3.3332243 -1.1781657 -1.2221891  1.2069107  0.1505145

$RowA
[1] 2

$ColA
[1] 4

$RowB
[1] 4

$ColB
[1] 3

$RowAB
[1] 2

$ColAB
[1] 3

[1] "list"

#### We can wrap it in an R function as well.

In [88]:
mmf <- function(A, B) {
    ra <- nrow(A)
    ca <- ncol(A)
    rb <- nrow(B)
    cb <- ncol(B)
    
    if (ca == rb) {
        result <- .Fortran("mat_mul_for",
                            A = as.double(A),
                            B = as.double(B),
                            AB = double(ra * cb),
                            RowA = as.integer(ra),
                            ColA = as.integer(ca),
                            RowB = as.integer(rb),
                            ColB = as.integer(cb),
                            RowAB = as.integer(ra),
                            ColAB = as.integer(cb)
                            )
        mm <- matrix(result$AB, nrow = result$RowAB, byrow = F)
    } else {
        stop('non-conformable arguments')
    }
    return(list(Result = mm,
                Dimension = c(result$RowAB, result$ColAB)
               )
          )
}

In [89]:
set.seed(20190813)

ra <- 2
ca <- 50000
rb <- 50000
cb <- 3

A <- matrix(rnorm(ra*ca), nrow = ra)
B <- matrix(rnorm(rb*cb), nrow = rb)

mmf(A, B)

$Result
          [,1]     [,2]       [,3]
[1,] -345.3068 359.6366  -54.33261
[2,] -190.4709  85.7216 -330.53902

$Dimension
[1] 2 3

In [90]:
A %*% B

[,1]      [,2]     [,3]      
[1,] -345.3068 359.6366  -54.33261
[2,] -190.4709  85.7216 -330.53902

<a id='4'></a>
## 4. [Exercises](#0)

### 4.1 Efficient coding
A fake dataset is generated. Results should make no biological sense.

In [91]:
set.seed(20200306)
N <- 200
height <- round(rnorm(n = N, mean = 180, sd = 10)) # in centimeter
weight <- round(rnorm(n = N, mean = 80, sd = 10)) # in kilograms
age <- round(rnorm(n = N, mean = 50, sd = 10))
treatment <- sample(c(TRUE, FALSE), size = N, replace = T, prob = c(0.3,0.7))
HF <- sample(c(TRUE, FALSE), size = N, replace = T, prob = c(0.1,0.9))

fake <- data.frame(height, weight, age, treatment, HF)
head(fake)

height weight age treatment HF   
1 186     92    60  FALSE     FALSE
2 155     74    58  FALSE      TRUE
3 182     79    62  FALSE     FALSE
4 178    101    54  FALSE     FALSE
5 182     72    54  FALSE     FALSE
6 159     66    41  FALSE      TRUE

1. (Vectorized operation) Calculate BMI for every individual ($\mathrm{BMI} = \mathrm{weight}(kg)/\mathrm{height}(m)^2$)

2. (Categorization) BMI Categories: 
    - Underweight = <18.5
    - Normal weight = 18.5–24.9 
    - Overweight = 25–29.9 
    - Obesity = BMI of 30 or greater

3. (*apply) Mean BMI of each BMI group

4. (Aggregation) Proportion with heart failure in each BMI-treatment group

In [92]:
# Trick:
FALSE+TRUE+TRUE

[1] 2

5. Write a function that allow user to specify
    - a dataset
    - the (binary) treatment variable
    - the (binary) outcome variable
    
 and return a cross-tabulation (a 2x2 table).

5 pro. The function should be able to check whether the treatment/outcome variables are binary or not. Continuous variables will be dichotomized based on a user-defined threshold.

### 4.2 Efficient computing
Make a function in R using R and/or Rcpp and/or R call Fortran that does


__Level 1__ Integer division of two integers using a loop

        I have 9 dollars to buy donuts for my colleagues. The donuts are 2 dollars each.
$$ 9>2 \rightarrow 9-2 = 7\quad \mathrm{1\, donut} $$
$$ 7>2 \rightarrow 7-2 = 5\quad \mathrm{2\, donuts}  $$
$$ 5>2 \rightarrow 5-2 = 3\quad \mathrm{3\, donuts}  $$
$$ 3>2 \rightarrow 3-2 = 1\quad \mathrm{4\, donuts}  $$
$$ 1<2 \rightarrow \mathrm{stop} $$

In [93]:
# Something like this.
9 %/% 2; 9%%2

[1] 4

[1] 1

__Level 2__ Element-wise integer division for two integer vectors

In [94]:
# Something like this.
c(15, 14, 13, 12) %/% c(6, 5, 4, 3)
c(15, 14, 13, 12) %% c(6, 5, 4, 3)

[1] 2 2 3 4

[1] 3 4 1 0

_If you like loops, loops in C++ and Fortran are fast._

__Level 3__ Linear regression

The formula for the point estimates is
$$ \boldsymbol{\beta} = (\boldsymbol{X}^\top \boldsymbol{X})^{-1} \boldsymbol{X}^\top \boldsymbol{Y} $$

- matrix transpose in R: t(X)
- matrix inverse in R: solve(X)
- matrix-matrix and matrix-vector mulplication in R: X %*% Y

In [95]:
# If you enter the right X and Y in your function, you should get the following result
lm(Wr.Hnd~NW.Hnd+Age, data = data)


Call:
lm(formula = Wr.Hnd ~ NW.Hnd + Age, data = data)

Coefficients:
(Intercept)       NW.Hnd          Age  
     1.1109       0.9535      -0.0103  


__Level 4__ Gradient descent to calculate the minimum value of a given function, with user-supplied gradient function.
- Gradient descent is an iterative algorithm therefore we have to use loops. Here we can really see the speed advantage of C++ and Fortran over R.

In [96]:
# Something like this, both inputs are R functions.
GD <- function(objective_function, gradient_function, initial_value) {
    statements
}

__Level 5__ Specific task in your own research.